unified-scheduler-logic/src/lib.rs

```diff
@@ -1,20 +1,1299 @@
+enum RequestedUsage {
+    Readonly,
+    Writable,
+}
+
+/// Internal scheduling data about a particular address.
+///
+/// Specifically, it holds the current [`Usage`] (or no usage with [`Usage::Unused`]) and which
+/// [`Task`]s are blocked to be executed after the current task is notified to be finished via
+/// [`::deschedule_task`](`SchedulingStateMachine::deschedule_task`)
+#[derive(Debug)]
+struct UsageQueueInner {
+    current_usage: Usage,
+    blocked_usages_from_tasks: VecDeque<UsageFromTask>,
+}
+
+type UsageFromTask = (RequestedUsage, Task);
+
+impl Default for UsageQueueInner {
+    fn default() -> Self {
+        Self {
+            current_usage: Usage::default(),
+            // Capacity should be configurable to create with large capacity like 1024 inside the
+            // (multi-threaded) closures passed to create_task(). In this way, reallocs can be
+            // avoided happening in the scheduler thread. Also, this configurability is desired for
+            // unified-scheduler-logic's motto: separation of concerns (the pure logic should be
+            // sufficiently distanced from any some random knob's constants needed for messy
+            // reality for author's personal preference...).
+            //
+            // Note that large cap should be accompanied with proper scheduler cleaning after use,
+            // which should be handled by higher layers (i.e. scheduler pool).
+            blocked_usages_from_tasks: VecDeque::with_capacity(128),
+        }
+    }
+}
+
+impl UsageQueueInner {
+    fn push_blocked_usage_from_task(&mut self, usage_from_task: UsageFromTask) {
+        self.blocked_usages_from_tasks.push_back(usage_from_task);
+    }
+
+    #[must_use]
+    fn pop_unblocked_usage_from_task(&mut self) -> Option<UsageFromTask> {
+        self.blocked_usages_from_tasks.pop_front()
+    }
+
+    #[must_use]
+    fn pop_unblocked_readonly_usage_from_task(&mut self) -> Option<UsageFromTask> {
+        if matches!(
+            self.blocked_usages_from_tasks.front(),
+            Some((RequestedUsage::Readonly, _))
+        ) {
+            self.pop_unblocked_usage_from_task()
+        } else {
+            None
+        }
+    }
+
+    fn has_no_blocked_usage(&self) -> bool {
+        self.blocked_usages_from_tasks.is_empty()
+    }
+}
+
+const_assert_eq!(mem::size_of::<TokenCell<UsageQueueInner>>(), 40);
+
+/// Scheduler's internal data for each address ([`Pubkey`](`solana_sdk::pubkey::Pubkey`)). Very
+/// opaque wrapper type; no methods just with [`::clone()`](Clone::clone) and
+/// [`::default()`](Default::default).
+#[derive(Debug, Clone, Default)]
+pub struct UsageQueue(Arc<TokenCell<UsageQueueInner>>);
+const_assert_eq!(mem::size_of::<UsageQueue>(), 8);
+
+/// A high-level `struct`, managing the overall scheduling of [tasks](Task), to be used by
+/// `solana-unified-scheduler-pool`.
+pub struct SchedulingStateMachine {
+    unblocked_task_queue: VecDeque<Task>,
+    active_task_count: ShortCounter,
```

Comment:
hehe, i agree tastelessness of `active` here. hmm, `current_task_count` is a bit ambiguous. how about `in_progress_task_count`? (or `not_(yet)_handled_task_count`; don't like this much because of implication of being a counter of tasks which has been resolved not to process).

Reply:
I find in_progress as confusing, since it implies that something is progressing while blocked tasks technically aren't progressing. @apfitzge as the native speaker, wdyt would be the best term here?

Reply:
how about `pending_task_count`, or `managed_task_count`, then?

Reply:
... or `runnable_task_count`?